## Likelihood Ratio

# Intuition


- **Likelihood Ratio (LR)**: The ratio of the likelihood for two models

**FIGURE PLACEHOLDER:** ![Likelihood Intuition Cartoon](image_placeholder)

# Notations

Likelihood Ratio (LR) that compares two fully-specified (discrete) models is simply the ratio of the likelihood for the two models given data $x$: 

$$
LR(M_1,M_0):=\frac{L(M_1)}{L(M_0)} = \frac{p(x|M_1)}{p(x|M_0)}
$$

Large values of $LR(M_1,M_0)$ indicate that the data are much more probable under model $M_1$ than under model $M_0$, and so indicate support for $M_1$. Conversely, small values of LR indicate support for model $M_0$.

Note that for fully specified models, the likelihood ratio is also known as the **"Bayes Factor" (BF)**. In this case, they are just two different names for the same thing.



<!-- 

The Likelihood Ratio Test (LRT) is a statistical test used to compare the fit of two models: a **null model** and an **alternative model**. The null model usually represents a simpler hypothesis (e.g., no effect), and the alternative model represents a more complex hypothesis (e.g., an effect exists). 

The basic idea is to test whether the data provides enough evidence to reject the null hypothesis in favor of the alternative hypothesis. Note that the Likelihood Ratio Test (LRT) is primarily a frequentist approach to hypothesis testing, not Bayesian.


The Likelihood Ratio Test statistic is defined as:

$$
\Lambda = -2 \log \left( \frac{L(\text{null})}{L(\text{alternative})} \right)
$$

Where:
- $L(\text{null})$ is the likelihood of the data under the null model,
- $L(\text{alternative})$ is the likelihood of the data under the alternative model.

The test statistic $\Lambda$ is compared to a **chi-squared distribution** with degrees of freedom equal to the difference in the number of parameters between the two models.

### Bayes Factor (BF)
In the Bayesian framework, we approach model comparison slightly differently than the frequentist approach. Instead of using the likelihood ratio to compute a p-value for hypothesis testing, Bayesian methods typically use the **Bayes Factor** (BF) to compare the models.

The **Bayes Factor** compares the marginal likelihoods (or evidence) of the models:

$$
\text{BF} = \frac{P(\text{data} | \text{model 1})}{P(\text{data} | \text{model 2})}
$$

Where:
- $P(\text{data} | \text{model 1})$ and $P(\text{data} | \text{model 2})$ are the marginal likelihoods (or evidence) of the models. In the frequentist LRT, this would correspond to the likelihoods of the two models.

In this way, the Bayes Factor provides a measure of the relative evidence for one model over another, rather than using a p-value to decide whether to reject a model.

 -->

# Example

Here we continue the example used in the `6_1_likelihood.ipynb` notebook.

In [13]:
set.seed(927)
# Simulate genotypes (0, 1, 2)
n <- 3  # number of individuals
genotypes <- sample(0:2, n, replace = TRUE)

# Simulate true mean and effect size
mu_0 <- 0  # population mean of the trait (e.g., height in cm) under the null model
beta_true <- 1  # true effect size of the genetic variant
sigma <- 0.1  # standard deviation of the trait

# Simulate trait values based on the alternative model
trait_values <- rnorm(n, mean = mu_0 + beta_true * genotypes, sd = sigma)

# Data preview
data.frame(genotype = genotypes, trait = trait_values)

genotype trait     
1 2         1.8983190
2 2         2.0308952
3 0        -0.1040512

In [14]:
# Likelihood for a model with mu = mu_0, sigma = sigma
likelihood <- function(mu, sigma, y) {
  n <- length(y)
  (1 / (sqrt(2 * pi) * sigma))^n * exp(-sum((y - mu)^2) / (2 * sigma^2))
  prod(dnorm(y, mean = mu, sd = sigma, log = FALSE))
}
likelihood_1 = likelihood(mu = mu_0, sigma, trait_values)
likelihood_2 = likelihood(mu = mu_0 + beta_true, sigma, trait_values)

cat("Likelihood of Model 1 (no causal effect):", likelihood_1, "\n")
cat("Likelihood of Model 2 (causal effect):", likelihood_2, "\n")

Likelihood of Model 1 (no causal effect): 5.662115e-167 
Likelihood of Model 2 (causal effect): 5.414108e-66 


The likelihood ratio comparing $M_2$ vs $M_1$ is therefore:

In [15]:
likelihood_2/likelihood_1

[1] 9.561988e+100

In [16]:
# Log-likelihood for a model with mu and sigma
log_likelihood <- function(mu, sigma, y) {
  # n <- length(y)
  # -0.5 * n * log(2 * pi * sigma^2) - sum((y - mu_0)^2) / (2 * sigma^2)
  sum(dnorm(y, mean = mu, sd = sigma, log = TRUE))
}
loglikelihood_1 = log_likelihood(mu = mu_0, sigma, trait_values)
loglikelihood_2 = log_likelihood(mu = mu_0+beta_true, sigma, trait_values)
cat("Log-likelihood of Model 1 (no causal effect):", loglikelihood_1, "\n")
cat("Log-likelihood of Model 2 (causal effect):", loglikelihood_2, "\n")

Log-likelihood of Model 1 (no causal effect): -382.7979 
Log-likelihood of Model 2 (causal effect): -150.2816 


The log-likelihood ratio comparing $M_2$ vs $M_1$ is therefore:

In [17]:
loglikelihood_2-loglikelihood_1

[1] 232.5163

The ratio of likelihood  (after taking the log) is equivalent to the difference of the log-likehood, and the $l$ is a monotonic function of $L$.

In [18]:
exp(log_likelihood_2 - log_likelihood_1)

[1] 9.561988e+100

# Summary
- The **"likelihood ratio"** that compares two fully-specified (discrete) models is simply the ratio of the likelihood for the two models given data $x$: 

$$
LR(M_1,M_0):=\frac{L(M_1)}{L(M_0)} = \frac{p(x|M_1)}{p(x|M_0)}
$$

where $L(M)$ denotes the likelihood for model $M$ under data $x$.

- The **"log-likelihood ratio"** is defined as the logarithm of the likelihood ratio:

$$
\log LR(M_1, M_0) := \log \frac{L(M_1)}{L(M_0)} = \log p(x | M_1) - \log p(x | M_0)
$$

where $l(M)$ denotes the log-likelihood for model $M$ under data $x$. The log-likelihood ratio measures the relative support for model $M_1$ over model $M_0$ on a **logarithmic scale**.

- A natural question is,  **what value of the LR (or LLR) should we treat as "convincing" evidence for one model vs another**? The short answer here is that this is **context dependent**, which depends on the relative plausibility of the models we are comparing. For more discussions, refer to this [notes](https://stephens999.github.io/fiveMinuteStats/LR_and_BF.html) by Matthew Stephens.

# Supplementary Materials

## Example 1 -- allele frequency

Here we use the same example 1 in the `6_1_likelihood.ipynb` notebook, and we have calculated the likelihood for the two models $M_S$ and $M_F$.

### Calculate LR

In [19]:
x = c(1,0,1,0,0,1)

fS = c(0.40, 0.12,0.21,0.12,0.02,0.32)
fF = c(0.8,0.2,0.11,0.17,0.23,0.25)

L = function(f,x){ prod(f^x*(1-f)^(1-x)) }

In [20]:
L(fS,x)
L(fF,x)

[1] 0.02039955

[1] 0.01124816

The **likelihood ratio (LR)** is simply the ratio of those two values:

In [21]:
L(fS,x)/L(fF,x)

[1] 1.81359

### How to interpret LR?

> So $LR(M_S,M_F;x)$ is 1.8135904. This means that the data favor the tusk coming from a savanna elephant by a factor of about 1.8. This is a fairly modest factor – not large enough to draw a convincing conclusion. We will have more to say about interpreting LRs, and what values might be considered “convincing” later.
>
> Note that we have deliberately focused on the likelihood ratio, and not the actual likelihood values themselves. This is because actual likelihood values are generally not useful - it is only the ratios that matter when comparing the models. One way of thinking about this is that the actual likelihood values are very context dependent, and so likelihoods from different data sets are not comparable with one another. However, the meaning of the likelihood ratio is in some sense consistent across contexts: LR =1.8 means that the data favour the first model by a factor of 1.8 whatever the context.

## Example 2 -- concertration of protein in blood

Here we use the same example 2 in the `6_1_likelihood.ipynb` notebook, and we have calculated the likelihood for the two models $M_0$ (normal individual group) and $M_1$ (diseased individual group).

### Calculate LR

In [22]:
X_val=4.02
# dgamma(x, shape, rate = 1, scale = 1/rate, log = FALSE) returns the Density for the Gamma distribution with parameters shape and scale at x
y0_val = dgamma(X_val,scale=0.5,shape=2)
y1_val = dgamma(X_val,scale=1,shape=2)
y0_val
y1_val

[1] 0.005182728

[1] 0.07217092

In [23]:
y1_val/y0_val

[1] 13.92528

### Interpretation

The $LR(M_1,M_0;x)$ is 13.9, i.e., the data favours the individual being diseased ($M_1$) over being normal ($M_0$) by a factor of approximately 14.



# Recommended Reading

- Section *Likelihood Ratio and Likehood* in [FiveMinuteStats](http://stephens999.github.io/fiveMinuteStats/index.html) by Matthew Stephens
